python/sglang/srt/layers/radix_attention.py

```diff
@@ -191,6 +207,257 @@ def decode_forward_flashinfer(self, q, k, v, input_metadata: InputMetadata):
 
+    def launch_sp_comm_ops(
+        self, kv_to_recv, kv_to_send, from_rank, to_rank, my_rank, sp_size, itr
+    ):
+        # Interleaving workers for send and recv to avoid deadlock
+        def _send_first():
+            flags = [None for _ in range(sp_size)]
+            for _rank in range(sp_size):
+                _next = _rank
+                flag = True
+                while flags[_next] is None:
+                    flags[_next] = flag
+                    _next = (_next + itr) % sp_size
+                    flag = not flag
+            return flags[my_rank]
+
+        def _send(handles, group):
+            if my_rank != to_rank:
+                to_global_rank = group.first_rank + to_rank
+                for t in kv_to_send:
+                    handles.append(
+                        P2POp(
+                            op=isend,
+                            tensor=t,
+                            peer=to_global_rank,
+                            group=group.device_group,
+                        )
+                    )
+
+        def _recv(handles, group):
+            if my_rank != from_rank:
+                from_global_rank = group.first_rank + from_rank
+                for t in kv_to_recv:
+                    handles.append(
+                        P2POp(
+                            op=irecv,
+                            tensor=t,
+                            peer=from_global_rank,
+                            group=group.device_group,
+                        )
+                    )
+
+        handles = []
+        reqs = []
+        sp_group = get_sp_group()
+
+        if _send_first():
+            _send(handles, sp_group)
+            _recv(handles, sp_group)
+        else:
+            _recv(handles, sp_group)
+            _send(handles, sp_group)
+        if handles:
+            reqs = batch_isend_irecv(handles)
+        return reqs
+
+    def wait_sp_comm_ops(self, reqs):
+        for req in reqs:
+            req.wait()
+
+    def seq_parallel_extend_forward_flashinfer(
+        self, q, k, v, input_metadata: InputMetadata
+    ):
+        """Here we adopted a unique parallelization strategy.
+        For each SP worker, we have either (1) QKV of entire sequences:
+            q tensor: [padded_total_num_tokens, q_head_num // SP_SIZE, head_dim]
+            k tensor: [padded_total_num_tokens, k_head_num, head_dim]
+            v tensor: [padded_total_num_tokens, v_head_num, head_dim]
+        Or (2) Q of entire sequences and KV of the current SP shard:
+            q tensor: [padded_total_num_tokens, q_head_num // SP_SIZE, head_dim]
+            k tensor: [padded_sp_shard_num_tokens, k_head_num, head_dim]
+            v tensor: [padded_sp_shard_num_tokens, v_head_num, head_dim]
+
+        Case (1) saves cross-SP-worker communication, while case (2) saves computation
+        to get K and V for entire sequences but need computation in SP attn.
+        """
```

Comment:
Just curious, does anyone have opinions on [TreeAttention](https://openreview.net/forum?id=jMZglnlwf7), which seem to be optimized for decoding?